liquid-rust-core/src/diagnostics.rs

```diff
@@ -0,0 +1,177 @@
+use liquid_rust_common::errors::ErrorReported;
+use rustc_session::{Session, SessionDiagnostic};
+
+pub(crate) struct Diagnostics<'tcx> {
+    sess: &'tcx Session,
+    errors: usize,
+}
+
+impl Diagnostics<'_> {
+    pub fn new(sess: &Session) -> Diagnostics {
+        Diagnostics { sess, errors: 0 }
+    }
+
+    pub fn raise<T>(&mut self) -> Result<T, ErrorReported> {
+        assert!(self.errors > 0);
+        self.errors = 0;
+        Err(ErrorReported)
+    }
+
+    // fn raise_if_errors(&mut self) -> Result<(), ErrorReported> {
+    //     if self.errors > 0 {
+    //         self.errors = 0;
+    //         Err(ErrorReported)
+    //     } else {
+    //         Ok(())
+    //     }
+    // }
+
+    pub fn emit_err<'a>(&'a mut self, err: impl SessionDiagnostic<'a>) -> &mut Self {
+        self.sess.emit_err(err);
+        self.errors += 1;
+        self
+    }
+}
+
+impl Drop for Diagnostics<'_> {
+    fn drop(&mut self) {
+        assert!(self.errors == 0);
+    }
+}
+
+pub(crate) mod errors {
```

Comment:
TODO

These are not general diagnostics. They should go to the final resolve module once we remove the old one. I think some of them should go to desugar.